### PR TITLE
core: Add `Array` template class to abstract a fixed-size contiguous buffer whose size is determined at runtime.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

In the current CLP, we use unique pointers of C-style arrays across the code base whenever we need a fix-sized buffer. This violates `clang-tidy` suggestions which advocate deprecating C-style arrays. However, existing `std` containers have limitations to use.
- For `std::vector`, it is hard to control the underlying storage. The storage size is controlled by its capacity but not the size. Also, the vector can't be made immutable to use the storage as a buffer to receive data (such as the use case in `NetworkReader`). This means there's no way to avoid reallocation/resizing by mistakes. Declaring a vector and using its underlying `data()` buffer will not provide any safety.
- For `std::array`, the size must be given in the compile time.

Therefore, we introduce our own implementation of arrays, `clp::Array`, which is a fix-sized array where the size is determined during the run-time. We use `std::unique_ptr` to manage the underlying buffer storage (which is allocated in the heap), and a member to keep track of the size to avoid out-of-bound access. This PR implements the basic features of our array, including methods for random access and iterators. At this stage, the underlying element of the array must be default initializable so that we don't have to implement the initializer list constructor yet.
APIs of this class are designed to mirror `std` containers.

# Validation performed

- Ensure all workflows can be built.
- Add unit tests to cover the basic usages against fundamental types and default constructible aggregated types, and ensure the iterator is compatible with `std` algorithms.

### CODE REVIEW

File: components/core/src/clp/Array.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+#include <type_traits>
+#include <utility>
+
+#include "ErrorCode.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+/**
+ * Class for a run-time fix-sized array.
+ * @tparam T The type of the element in the array. The type must be default constructable.
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:17]:
```suggestion
 * Class for a runtime fix-sized array.
 * @tparam T The type of elements in the array. The type must be default initializable so that the
 * class doesn't need to implement an initializer list constructor.
```

Is that accurate?

File: components/core/src/clp/Array.hpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ */
+template <typename T>
+requires(std::is_fundamental_v<T> || std::is_default_constructible_v<T>)
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:20]:
```suggestion
requires(std::is_fundamental_v<T> || std::is_default_initializable<T>)
```

[This](https://stackoverflow.com/questions/62577956/what-is-the-difference-between-stddefault-initializable-and-stdis-default) seems to suggest `std::is_default_initializable` is stricter for our requirements?

File: components/core/src/clp/Array.hpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    [[nodiscard]] auto empty() const -> bool { return 0 == size(); }
+
+    /**
+     * @return The size of the array.
+     */
+    [[nodiscard]] auto size() const -> size_t { return m_size; }
+
+    /**
+     * @return The ptr of the underlying data buffer.
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:77]:
```suggestion
     * @return A pointer to the underlying data buffer.
```

File: components/core/src/clp/Array.hpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    [[nodiscard]] auto data() -> T* { return m_data.get(); }
+
+    /**
+     * @return The ptr of the underlying data buffer.
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:82]:
```suggestion
     * @return A pointer to the underlying data buffer.
```

File: components/core/src/clp/Array.hpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    [[nodiscard]] auto begin() const -> ConstIterator { return m_data.get(); }
+
+    [[nodiscard]] auto end() const -> ConstIterator { return m_data.get() + m_size; }
+
+private:
+    /**
+     * @param idx
+     * @throw `OperationFailed` if the given index is out of bound.
+     */
+    auto assert_idx(size_t idx) -> void {
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:119]: How about `assert_is_in_range`?

File: components/core/src/clp/Array.hpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        if (idx >= m_size) {
+            throw OperationFailed(
+                    ErrorCode_OutOfBounds,
+                    __FILE__,
+                    __LINE__,
+                    "`clp::Array access out of bound.`"
+            );
```
Comment (kirkrodrigues) [inline components/core/src/clp/Array.hpp:126]: How about using `std::out_of_range` instead?

* `std::out_of_range` seems like the standard exception to throw when an out of bounds access is made.
* `TraceableException` doesn't seem to offer much benefit in this case since there's only one place where the exception could be thrown from and the message indicates as much.

File: components/core/tests/test-Array.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    for (auto const val : clp_array) {
+        REQUIRE((val == clp_array.at(idx)));
+        REQUIRE((val == std_vector.at(idx)));
+        ++idx;
+    }
+    REQUIRE((cBufferSize == idx));
+    REQUIRE_THROWS(clp_array.at(idx));
+}
+
+TEST_CASE("array_default_constructable", "[clp::Array]") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-Array.cpp:40]:
```suggestion
TEST_CASE("array_default_initializable", "[clp::Array]") {
```

File: components/core/tests/test-Array.cpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    vector<std::string> const std_vector{"yscope", "clp", "clp::Array", "default_constructable"};
```
Comment (kirkrodrigues) [inline components/core/tests/test-Array.cpp:41]:
```suggestion
    vector<std::string> const std_vector{"yscope", "clp", "clp::Array", "default_initializable"};
```

File: components/core/tests/test-Array.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <Catch2/single_include/catch2/catch.hpp>
+
+#include "../src/clp/Array.hpp"
+
+using clp::Array;
+using std::vector;
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("array_fundamental", "[clp::Array]") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-Array.cpp:14]: Let's also test creating an empty array.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp/Array.hpp (hunk 1 above)
Comment (LinZhihao-723) [reply components/core/src/clp/Array.hpp:17]: yes, that's more accurate. Thanks!

File: components/core/src/clp/Array.hpp (hunk 2 above)
Comment (LinZhihao-723) [reply components/core/src/clp/Array.hpp]: Make sense

File: components/core/src/clp/Array.hpp (hunk 6 above)
Comment (LinZhihao-723) [reply components/core/src/clp/Array.hpp:127]: True

File: components/core/src/clp/Array.hpp (hunk 2 above)
Comment (LinZhihao-723) [reply components/core/src/clp/Array.hpp]: btw it's `[std::default_initializable](https://en.cppreference.com/w/cpp/concepts/default_initializable)`

Comment (kirkrodrigues) [reply components/core/src/clp/Array.hpp]: Right, my mistake. 5am, lol.

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

core: Add `Array` template class to abstract a fixed-size contiguous buffer whose size is determined at runtime.

**Comment by kirkrodrigues** [comment]
Workflow failure is unrelated and is being tracked in #519.
